classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php

```diff
@@ -168,4 +188,235 @@ private function _makeRemoteUrlLocalizable() {
 			$table->dropColumn('url');
 		});
 	}
+
+	/**
+	 * Migrate submission files after major refactor
+	 *
+	 *	- Add files table to manage underlying file storage
+	 *	- Replace the use of file_id/revision as a unique id with a single
+	 * 		auto-incrementing submission_file_id, and update all references.
+	 *	- Move revisions to a submission_file_revisons table.
+	 *	- Drop unused columns in submission_files table.
+	 *
+	 * @see pkp/pkp-lib#6057
+	 */
+	private function _migrateSubmissionFiles() {
+		import('lib.pkp.classes.submission.SubmissionFile'); // SUBMISSION_FILE_ constants
+
+		// Create a new table to track files in file storage
+		Capsule::schema()->create('files', function (Blueprint $table) {
+			$table->bigInteger('file_id')->autoIncrement();
+			$table->string('path', 255);
+		});
+
+		// Create a new table to track submission file revisions
+		Capsule::schema()->create('submission_file_revisions', function (Blueprint $table) {
+			$table->bigInteger('revision_id')->autoIncrement();
+			$table->bigInteger('submission_file_id');
+			$table->bigInteger('file_id');
+		});
+
+		// Add columns to submission_files table
+		Capsule::schema()->table('submission_files', function (Blueprint $table) {
+			$table->bigInteger('new_file_id'); // Renamed at the end of the migration
+		});
+
+		// Drop unique keys that will cause trouble while we're migrating
+		Capsule::schema()->table('review_round_files', function (Blueprint $table) {
+			$table->dropUnique('review_round_files_pkey');
+		});
+
+		// Create entry in files and revisions tables for every submission_file
+		import('lib.pkp.classes.file.FileManager');
+		$fileManager = new FileManager();
+		$rows = Capsule::table('submission_files')
+			->orderBy('file_id')
+			->orderBy('revision')
+			->get([
+				'file_id',
+				'revision',
+				'submission_id',
+				'genre_id',
+				'file_stage',
+				'date_uploaded',
+				'original_file_name'
+			]);
+		foreach ($rows as $row) {
+			// Reproduces the removed method SubmissionFile::_generateFileName()
+			// genre is %s because it can be blank with review attachments
+			$filename = sprintf(
+				'%d-%s-%d-%d-%d-%s.%s',
+				$row->submission_id,
+				$row->genre_id,
+				$row->file_id,
+				$row->revision,
+				$row->file_stage,
+				date('Ymd', strtotime($row->date_uploaded)),
+				strtolower_codesafe($fileManager->parseFileExtension($row->original_file_name))
+			);
+			$contextId = Capsule::table('submissions')->where('submission_id', '=', $row->submission_id)->first()->context_id;
+			$path = sprintf(
+				'%s/%s/%s',
+				Services::get('submissionFile')->getSubmissionDir($contextId, $row->submission_id),
+				$this->_fileStageToPath($row->file_stage),
+				$filename
+			);
+			if (!Services::get('file')->fs->has($path)) {
+				throw new Exception("A submission file was expected but not found at $path.");
```

Comment:
This will happen in practice during upgrades, and I suggest treating it as a warning and continuing rather than stopping the upgrade dead in the water until the user puts a dummy file there, restores from backup, and re-runs.

Reply:
What's the correct way to print a warning? I'm seeing some `error_log` calls in the upgrade scripts. Is that it?

Reply:
I'm trying to think through the implications here if the upgrade script skips the rest of this loop due to a missing file. I'm not sure, but I think that this could have a knock-on effect where event logs point to the wrong file id.

Would it maybe be better here to set the `$newFileId` to `0`? I think that might cause some havoc too, but at least it won't point to an incorrect file id?

Reply:
Yes, warnings should use `error_log`. I'd recommend using a value of `NULL` rather than `0` for missing data -- you may need to permit NULLs in that column if they aren't, and if so, please make a note why that's the case. `0` IDs won't be permitted any more once we start using proper `FOREIGN KEY` constraints.